prow/sidecar/run.go

```diff
@@ -125,7 +129,12 @@ func (o Options) Run(ctx context.Context) (int, error) {
 				// second upload but we can tolerate this as we'd rather get SOME
 				// data into GCS than attempt to cancel these uploads and get none.
 				logrus.Errorf("Received an interrupt: %s, cancelling...", s)
-				cancel()
+
+				err = o.doUpload(spec, false, true, metadata, buildLogs)
```

Comment:
We need this preemptive upload to be interrupted if we see that the marker files are written and we're able to begin the main upload.
To achieve this we should instrument the call stack of `doUpload` with a context that can be used to cancel the upload early and then pass the `ctx` context here. That way if the `wait` function returns on line 148 we'll cancel the preemptive upload on line 150 and begin the main upload.

Reply:
Handled this by adding a new method `doBestEffortUpload` which uses `doUpload` but also switches context when main upload is ready.